### PR TITLE
Move common graphics utils to //flutter/common/graphics

### PR DESCRIPTION
**Pull request by iskakaushik**
Move common graphics utils to //flutter/common/graphics

    This change creates a new source set //flutter/common/graphics
    with minimal dependencies.

    The intention with this target is to be later consumed in a
    darwin common graphics target across the desktop boundary.

    Also unifies some of the graphics utilities under the same folder.

### CODE REVIEW

**Comment by iskakaushik** [comment]
@stuartmorgan , I eventually plan on creating `darwin/graphics` like so: https://github.com/flutter/engine/pull/22303/files#diff-b8669fd3d05e9e8799cf1dd1b38549f32f06b79ab6b7ae326aeb8dc97bca9d4a. This will let us share a bunch of metal logic for example across ios/mac boundaries. And then use this target as a dependency for the mac desktop embedding.

File: graphics/BUILD.gn
Comment (stuartmorgan) [inline graphics/BUILD.gn:24]: Shouldn't this have `public_configs = [ "//flutter:config" ]` like most other targets?

File: graphics/persistent_cache.cc
Comment (stuartmorgan) [inline graphics/persistent_cache.cc:5]: I'm not wild about having flutter/graphics have a dependency into flutter/shell (for version). I know version is essentially self-contained, but shell/* is exactly what we don't want embeddings built on embedder.h to have arbitrary dependencies into. That's the whole reason for this move, after all.

I think we need a clearer definition of what dependencies are allowable, and a directory structure that makes the line much crisper. (I really wish we had Chromium's tooling for whitelisting/blacklisting cross-directory dependencies set up.)

File: graphics/BUILD.gn
Comment (stuartmorgan) [inline graphics/BUILD.gn:18]: Since we don't have any enforcement mechanism, we should at least have a big obvious comment here communicating that arbitrary dependencies should *not* be added here.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: graphics/persistent_cache.cc
Comment (iskakaushik) [reply graphics/persistent_cache.cc:5]: I agree, I was contemplating moving `version` out of shell to flutter root. Do you think that would make things easier to reason about?

File: graphics/BUILD.gn
Comment (iskakaushik) [reply graphics/BUILD.gn:18]: Makes sense. Will do.

**Review by liyuqian — COMMENTED** [review]
Moving `gl_context_switch, texture, persistent_cache` into `graphics` makes sense to me. @flar and I were also discussing where to put common graphics code shared between `flow` and `lib/ui`. I guess `graphics` would be a nice place once this lands.

File: graphics/persistent_cache.cc
Comment (stuartmorgan) [reply graphics/persistent_cache.cc:5]: I'm not sure how many things we want to move to the root; these two things are probably not going to be the last things we want to share.

I'm increasingly thinking we should instead have a new directory in shell, something like `shell/embedding_utils` (it's not a great name, but naming it is non-trivial, especially given the existing `common`, which now isn't), and make that the firewall, and put `graphics` and `version` in there, along with an explanatory README.

I'll workshop a name in the engine channel.

Comment (iskakaushik) [reply graphics/persistent_cache.cc:5]: @stuartmorgan given that this will likely take some thinking. I'd like to not let this PR go stale. We can always move this folder to the right home once we decide on it. I've also added a warning message in the gn file to not add additional deps.

Comment (stuartmorgan) [reply graphics/persistent_cache.cc:5]: My concern is that this gives the appearance of being isolated from shell/ by moving out of shell, but since it actually does depend back into shell that's not actually the case.

Even if we don't think this is a final location, what's the argument for having the interim location be both top-level and having dependencies into shell?

Comment (iskakaushik) [reply graphics/persistent_cache.cc:5]: I will move this into `//flutter/shell/graphics/`

Comment (liyuqian) [reply graphics/persistent_cache.cc:5]: What do you think of `//flutter/common/graphics` so we don't inflate the root? Maybe `version` can also go there as both of them don't seem to rely on `shell` to exist.

**Comment by iskakaushik** [comment]
@stuartmorgan , @liyuqian I've moved it to `//flutter/shell/common/graphics`. I will move `version` to `shell/common` in a follow-up commit.

File: common/graphics/persistent_cache.h
Comment (stuartmorgan) [inline common/graphics/persistent_cache.h:5]: All the header guards need updating again.

**Review by stuartmorgan — APPROVED** [review]
LGTM once the header guards are fixed.

**Review by liyuqian — APPROVED** [review]
LGTM. BTW, some `include` needs to be reordered. For example
```
-#include "flutter/fml/logging.h"
 #include "flutter/common/graphics/persistent_cache.h"
+#include "flutter/fml/logging.h"
```

**Pull request merged**
